abdiff/core/build_ab_images.py

```diff
@@ -1 +1,92 @@
-"""abdiff.core.build_ab_images"""
+import logging
+import tempfile
+
+import docker
+import docker.models
+import docker.models.images
+from pygit2 import clone_repository
+from pygit2.enums import ResetMode
+
+from abdiff.core.utils import update_or_create_job_json
+
+logger = logging.getLogger(__name__)
+
+
+def build_ab_images(
+    job_directory: str,
+    commit_sha_a: str,
+    commit_sha_b: str,
+    docker_client: docker.client.DockerClient | None = None,
+) -> tuple[str, str]:
+    """Build Docker images based on 2 commit SHAs.
+
+    Args:
+        job_directory: The directory containing all files related to a job.
+        commit_sha_a: The SHA of the first commit for comparison.
+        commit_sha_b: The SHA of the second commit for comparison.
+        docker_client: A configured Docker client.
+    """
+    if not docker_client:
+        docker_client = docker.from_env()
+    image_names = []
+    for commit_sha in [commit_sha_a, commit_sha_b]:
+        logger.debug(f"Processing commit: {commit_sha}")
+        image_name = f"transmogrifier-{job_directory.split("/")[-1]}-{commit_sha}:latest"
+        if docker_image_exists(docker_client, image_name):
+            image_names.append(image_name)
+        else:
+            image = clone_repo_and_build_image(job_directory, commit_sha, docker_client)
+            image_names.append(image.tags[0])
+        logger.debug(f"Finished processing commit: {commit_sha}")
+    images_data = {"image_name_a": image_names[0], "image_name_b": image_names[1]}
+    update_or_create_job_json(job_directory, images_data)
+    return (image_names[0], image_names[1])
+
+
+def docker_image_exists(
+    docker_client: docker.client.DockerClient, image_name: str
+) -> bool:
+    """Check if Docker image already exists with a certain name.
+
+    Args:
+        docker_client: A configured Docker client.
+        image_name: The name of the Docker image to be created.
+    """
+    return image_name in docker_client.images.list()
```

Comment:
Update: after some local testing, I'm not sure this `image_name in docker_client.images.list()` works as expected.

The output I'm seeing from `.images.list()` looks like they are `Image` instances:
```
[<Image: 'transmogrifier-hello-world-1910426fbcc98e098de222ccdcce3cf3ac1337e7:latest'>,
 <Image: 'transmogrifier-hello-world-395e612f8887b279ccf35363694bafb06ef9a25a:latest'>,
 <Image: ''>,
 <Image: ''>,
 <Image: 'test-job-395e612:latest'>,
 <Image: 'test-job-008e20c:latest'>,
...
...
]
```

But we're passing a string, so it seems possible the `in` operator doesn't find a string in a list of `Image` instances.

Wonder if this needs to not only get the images, but also then grab the first (or all?) image tags? 

Example:
```python
image = docker_client.images.list()[0]

image.tags[0]
# Out[13]: 'transmogrifier-hello-world-1910426fbcc98e098de222ccdcce3cf3ac1337e7:latest'
# this look like what we'd want to match on...
```

Reply:
I did manually test this but must have botched it, you are correct and I've refactored!